_posts/2020-09-02-mutiny-redis.adoc

```diff
@@ -0,0 +1,139 @@
+
+[source, java]
+----
+Multi<String> keys = redis.keys("*")
+     .onItem().transformToMulti(array -> Multi.createFrom().iterable(array))
+     .onItem().castTo(String.class);
+----
+
+This snippet:
+
+1. retrieves the `JsonArray` containing the keys
+2. creates a `Multi` streaming these keys, it's a `Multi<Object>` as a `JsonArray` is extending `Iterable<Object>`
+3. maps the items from this `Multi` to `String`
+
+At this point, we have a stream of (String) keys. So, we are done with step 1. 
+
+Now, step 2: for each key, we want to retrieve the associated object.
+
+So let's use the `hgetall` method:
+
+[source, java]
+----
+Multi<JsonObject> objects = keys
+  .onItem().transformToUniAndMerge(key -> redis.hgetall(key));
+----
+
+This snippet requires a bit of an explanation. 
+
+For each item of the stream `keys`, we call `hgetall`, which produces a `Uni<JsonObject>`. 
+
+So, we want to transform our key into a Uni (*transformToUni*).
+
+When you have a stream of items and need to invoke an asynchronous action for each item, you must choose how you will merge the results. Mutiny provides two strategies:
+
+- *merge -* as soon as the item produced by the `Uni` is received we send it downstream
+- *concatenate* - we preserve the order of the input stream to be sure that the items are sent downstream in the same order
+
+Let's illustrate this. Imagine we have the keys `1`, `2`, `3` and to the stream `{1, 2, 3}`. Also, let's consider that in our Redis database, the key `1` is associated to `A`, `2` to `B` and `3` to `C`. 
+
+If you use the *merge* strategy, we are retrieving the associated objects we the order is undetermined. We can end up with `{A, C, B}` or `{B, A, C}`. It depends on many factors, such as the latency,  scheduling, load and so on. However, it also means we can retrieve all the associated objects concurrently and produce the resulting stream without taking care of the order.
+
+If you use the *concatenate* strategy, it preserves the order from the input stream. So, it will always produce `{A, B, C}`. While it may be desirable, it may reduce the ability to retrieve the object concurrently, as Mutiny has to *wait* for all the retrieval of all the previous objects. For example, if Mutiny receives `C` first, it needs to wait for `A` and `B` before sending `C` downstream. 
+
+In our context, let's not preserve the order and use the *merge* strategy. So we use `transformToUniAndMerge`. 
+
+If you run the code multiple times, you might see order changes in the resulting array.
+
+Ok, step 2 done. Let's focus on the final steps: accumulate the objects into a `JsonArray`, and produce a `Uni<JsonArray>`, containing all the objects. Mutiny provides methods to gather items from a stream into lists, maps, sets, but not `JsonArray`. Fortunately, it offers a method that let you do this:
```

Comment:
What's the "it" that offers a method that allows you to do it, as you just mentioned Mutiny doesn't provide a method

Reply:
Rephrased.